rfcs/converged-implementation-patterns.md

```diff
@@ -0,0 +1,171 @@
+
+Tries to commit the common component pattern to writing in the repo for future reference and track improvements
+
+Code samples included could possibly be applied to utility scripts to generate packages
+
+## Background
+
+<!-- If there is relevant background include it here -->
+
+Many new engineers can find it hard to adapt to a new codebase, especially when the coding patterns and practices are fairly new and in a constant state of flux
+
+## Problem statement
+
+<!--
+Why are we making this change? What problem are we solving? What do we expect to gain from this?
+
+This section is important as the motivation or problem statement is indepenent from the proposed change. Even if this RFC is not accepted this Motivation can be used for alternative solutions.
+
+In the end, please make sure to present a neutral Problem statement, rather than one that motivates a particular solution
+-->
+
+Many new developers might have questions on the patterns that we follow to write a React component. This RFC attempts to commit and possibbly improve our code patterns in writing so that there might be a reference for new (and old) members of the team.
+
+## Detailed Design or Proposal
+
+<!-- This is the bulk of the RFC. Explain the proposal or design in enough detail for the inteded audience to understand. -->
+
+### Hook based architecture.
+
+Generally a component will have the following different files. Let us consider a `Sample` component
+
+- Sample.tsx
+- useSample.ts
+- useSampleClasses/useSampleStyles.ts
+- renderSample.tsx
+- Sample.types.ts
+
+#### renderSample.tsx
+
+renders the correct JSX output of the component and its slots given the correct state.
+
+```typescript=
+export const renderSample = (state: SampleState) => {
+  const { slots, slotProps } = getSlots(state);
+
+  return (
+    <slots.root {...slotProps.root}>
+      <slot.otherSlot {...slotProps.otherSlot} />
+    </slots.root>
+  );
+};
+```
+
+#### useSample.ts
+
+Accepts the component props and handles and internal state that the component might need.
+
+`State` here can be pretty broad, you could also consume context or create effects. This hook should be what the component relies on to function/render
+
+```typescript=
+/**
+ * Defines the different slots that can be rendered in this component
+ *
+ * This means that an icon can be rendered 'somewhere'
+ */
+export const sampleShorthandProps = ['icon'];
+
+// Creates a helper function to merge props while respecting slot props
+const mergeProps = makeMergeProps({ deepMerge: sampleShorthandProps });
+
+/**
+ * @parameter props -> these are normal React props for JSX components
+ * @parameter ref -> In case someone wants a ref to the root DOM element
+ * @parameter defaultProps -> safe default props
+ */
+export const useSample = (props: SampleProps, ref: React.Ref<HTMLElement>, defaultProps?: SampleProps) => {
+  const resolvedRef = ref || React.useRef();
+  const { checkedValues, onCheckedValuesChange } = useMenuListContext();
```

Comment:
Ditto for `useMenuListContext`. IMHO if this RFC managed to be mostly self-contained, it would be the main point of entry for the broader community, and currently it's really close to being just that!

Reply:
whoops, I copy/pasted from actual examples, I need to correct all of these to my `Sample` component

Reply:
I _**think**_ I've removed all terminology not related to `Sample`